### ISSUE TITLE
R.keys doesn't retrieve char indexes in string primitives

### ISSUE THREAD

**Issue by ascartabelli** [ramda/ramda#1761]
While I'm struggling to decide whether to let exceptions being thrown for `null` and `undefined` values in my own library (any insight is welcome), I noticed that you have a bug in `R.keys` here: https://github.com/ramda/ramda/blob/v0.21.0/src/keys.js#L47 (and later in line 50) due to the `Object(obj) !== obj` check.

The check is fine for string objects, not for primitives:

``` javascript
R.keys("foo") // => []
R.keys(new String("foo")) // => ["0", "1", "2"]
```

As you have already decided of not throwing exceptions, this can be easily solved with `compose(Object.keys, Object)` on line 47 and maybe in a _quick'n'dirty_ way on line 50 by replacing the check with `obj = Object(obj)`.
That second bit is _quick'n'dirty_, but it should have less impact on performance than a more complicated check to allow an early exit from the function.

**Comment by CrossEye** [comment]
I'm not sure this is a bug.

Why do you think that `keys` should report the indices?

Note that the signature is essentially `:: Object -> [String]`.  The behavior on primitives is not defined, but I'm happy with returning an empty list in that case.  `new String('foo')` _is_ an Object, by definition, and will be expected to have different behavior.

**Comment by ascartabelli** [comment]
> Why do you think that keys should report the indices?

@CrossEye Well, generically speaking I'm not a fan of considering primitives and their wrappers as different things: always thought it leads to confusion and I prefer to have a uniform behaviour in my code.
Also because I consider strings to be array-like objects, and I noticed only now that you don't in `R.isArrayLike`.

Anyway if this is the intended behaviour I think you should consider changing `keysIn` accordingly for consistency, as it returns indexes for both string primitives and objects.

**Comment by ascartabelli** [comment]
> Also because I consider strings to be array-like objects, and I noticed only now that you don't in `R.isArrayLike`

@CrossEye Although I just noticed that:

```javascript
R.filter(c => ~"aeiou".indexOf(c), "hello") // => ["e", "o"]
R.map(R.toUpper, "hello") // => ["H", "E", "L", "L", "O"]
```

**Comment by CrossEye** [comment]
We absolutely do treat Strings sometimes as list-like.  But I wouldn't document that `keys(['a','b','c']); //=> ['0', '1', '2']` either.  It's simply an accident of the implementation, and the somewhat odd fact that JS builds its arrays atop its objects.  It doesn't make sense with our logical type of lists.  It makes no more sense with strings.

All this is my opinion, and I would love to hear how others think.  @buzzdecafe, @davidchambers, @scott-christopher, @asaf-romano, @TheLudd?

**Comment by davidchambers** [comment]
Since you asked, Scott, I'll state my preference. You could probably have guessed. ;)

```javascript
R.keys('foo');
// ! TypeError: Invalid value
//
//   keys :: Object -> Array String
//           ^^^^^^
//             1
//
//   1)  "foo" :: String
//
//   The value at position 1 is not a member of ‘Object’.
```

**Comment by ascartabelli** [comment]
Sorry, @CrossEye , I think I did a poor job in stating my point.
I used the word bug because at first glance I supposed that it wasn't the intended behaviour, but what really mattered to me was pointing out an inconsistency.

If I see `keysIn(“foo”)` returning an array of indexes I expect `keys(“foo”)` to do the same, without even starting the `”foo”` / `new String(“foo”)` debate.

Moreover I can map over a string, filter it, even reduce it but I have an `isArrayLike` function returning `false` on the very same value.

Some people think that things like that should fall in the “undocumented behaviour” category and left alone, given the type system in JavaScript, while I think that for the very same reason I should put in extra effort when releasing a public API.

I’m not trying to nitpick or criticize, mind me, I do think you all did a **really great** job with Ramda.

Well, maybe nitpicking a bit, but with the best of intentions; I was doing the very same job on my code and I started with what I hoped to be the easier task: consistent behaviour on when exceptions are raised.

Soon I was out, looking at other libraries, trying to understand if there’s some kind of general consensus or to infer the best course of action.

**Comment by CrossEye** [comment]
@ascartabelli:
Please don't hesitate to nitpick.  And certainly don't hesitate to criticize.  Anyone around here who doesn't have a thick skin better grow one soon.

I'm sure that I occasionally get defensive about decisions here -- I'm human -- but I don't think this is one of those times.  This simply highlights some of the fundamental decision points in the design of the library.  And there is still some controversy around these.

The response above from @davidchambers demonstrates one of these controversial issues.  Being much more strict about what our functions accepted, and throwing errors when users supplied parameters that didn't match our expectations would be one way to handle these sorts of of issues.  It would bring working with Ramda closer to working with statically typed languages, which seem to have the upper-hand in the FP world these days.  That comes at the cost of being less familiar and less inviting to people used to Javascript, though.  [Sanctuary][sa] was built on the other side of this divide: it's more strict about types, more willing to throw exceptions, probably significantly slower to run, less natural to Javascript developers, but much more familiar to those coming from statically typed languages such as Haskell, OCaml, Scala, or F#.  To my mind, it's a very good thing to have both libraries out there.  (And BTW, the communities of these two have a huge overlap.)

That `keysIn` and `keys` return different things should not be any surprise in general, since the whole point of the various `*In` functions is to add prototype properties to the own-properties of their base functions.  But this is different, because those are not actually prototype properties.  This to me speaks of issues with `keysIn` that I think we'll need to address: it probably should act like `keys` on primitives.  I'd like others to weigh in here, though, too.

> Moreover I can map over a string, filter it, even reduce it but I have an `isArrayLike` function returning `false` on the very same value.

These definitely are more in the "accidents of implementation" category.  `filter` returns a fairly useful result, perhaps, and perhaps one can usefully `reduce` over a string, but these will be rare, and the `map` that happens doesn't really fulfil the logical contract of `map`.

As to `isArrayLike`, I think that's just nonsense that we should remove.  I can't recall how we got it, and it has no place I can think of on our public API.  I know that @buzzdecafe has suggested this several times.  I'm in favor of that.  I have no clue at all why the implementation specifically rejects strings, as they seem as likely as several others to meet the definition supplied.

> Some people think that things like that should fall in the “undocumented behaviour” category and left alone, given the type system in JavaScript, while I think that for the very same reason I should put in extra effort when releasing a public API.

This is where things get tricky.  To my mind, I am writing a library for me.  I hope it works well for other people, and I will make some concessions to that, but it is not designed to work for what is best for a random developer.  It's designed to work well for me.  Now of course this isn't me alone; there are a number of other major players involved.  Both @buzzdecafe and @davidchambers have many more commits than I do and there are others with many major contributions.  But that just expands the group of people whose API ideas are guiding it.  It doesn't change the fundamental fact that this library is designed to for a certain style of development.  @buzzdecafe and I started this library as a way of learning how to do FP better in Javascript.  We were thunderstruck when it started becoming so popular.  But we've always tried to keep true to the notion of Ramda being a tool to help us work the way we would prefer to work.

The way I want to work is generally without a lot of hand-holding from my tools.  I want functions to do the right things when I give them appropriate data.  When I give them bad data, it's fine if the results are perhaps odd.  I would rather that than have them start throwing exceptions.  Exceptions do not work well with declarative pipelined code, and if I have to start coding for them, I will have to write much less ergonomic code.  So Ramda generally only throws exceptions when it cannot make any reasonable guess about what to do with the data supplied.

I am _not_ recommending that you make the same decision for your library.  But I would recommend that you think more about how you would want to use the library than how you imagine others might want to.  If you get one that works well for you, it's almost certain that there are others who will think the same way and will also find it useful.  The only other way I know of to build a successful library is to find someone else who's developed a successful API, copy that and add some sort of minor improvements to it and call it your own; that's a pretty boring path.



  [sa]: http://sanctuary.js.org/

**Comment by ascartabelli** [comment]
@CrossEye:

> Please don't hesitate to nitpick. And certainly don't hesitate to criticize.

Will do!

> I'm sure that I occasionally get defensive about decisions here -- I'm human -- but I don't think this is one of those times.

Oh, no, I didn't think you were defensive at all: I only tried to clarify my point of view.

> This to me speaks of issues with keysIn that I think we'll need to address: it probably should act like keys on primitives.

Always thumbs up when your goal is consistency.
Although let me say again that, while I really understand @davidchambers concerns and yours about types, I firmly believe that primitives and their object wrappers should be treated as the same thing.

> As to isArrayLike, I think that's just nonsense that we should remove.

My 2 cents: I think you should. 
Array-like is a concept, not a type and not even an interface. You can't do a check against a concept and for sure you can't do anything useful with such check; only introduce harmful behaviour.

That said, I acknowledge the presence of the concept in my library, even by writing "array-like" in doc comments' parameters, and I let array related functions run stating that you'll get an array back (even if it means breaking the concept of `map` as you pointed out).
User's choice: it's documented and I'm guilty.

> The response above from @davidchambers demonstrates one of these controversial issues. Being much more strict about what our functions accepted, and throwing errors when users supplied parameters that didn't match our expectations would be one way to handle these sorts of of issues.

[...]

>  Sanctuary was built on the other side of this divide

[...]

> The way I want to work is generally without a lot of hand-holding from my tools. I want functions to do the right things when I give them appropriate data. When I give them bad data, it's fine if the results are perhaps odd. I would rather that than have them start throwing exceptions.

Personally I wholeheartedly agree with you, but I understand Sanctuary's design decision and I have nothing against it.
One of the design goals of my library is not throwing exceptions at all explicitly: I have to be able to code "letting things flow" and let the language itself be the one throwing them.
Suits my library's somewhat narrower scope, as Sanctuary's design suits its own.

The doubt that brought me here in the first place was deciding what to do in general with "nil" values: does having null safe functions hide too much?
i.e. ES6's `Object.keys` now throw exceptions only with "nil" values, Ramda chose to let them pass.

> I am not recommending that you make the same decision for your library. But I would recommend that you think more about how you would want to use the library than how you imagine others might want to.

I understand what you're suggesting, but I'm not at risk here: my train of thoughts usually goes elsewhere.
It starts from need, sure, and when I said:

>> I should put in extra effort when releasing a public API.

I didn't mean switching my focus blindly to an hypothetical user's needs.

But this would be long to discuss, and I'm already sorry for transforming an issue tracker into a forum.

We don't mutate things.

Thanks a lot for your time and your insights.

**Comment by CrossEye** [comment]
@ascartabelli:

>> This to me speaks of issues with keysIn that I think we'll need to address: it probably should act like keys on primitives.

> Always thumbs up when your goal is consistency.

Ok, but I'm afraid my reason for this change is different from yours.  For me it's simply that `keys`/`keysIn` is about the names of object properties.  And a string does not have properties named '0', '1', '2', etc, no more than a list does.  The facts that the language implements arrays atop objects and that it implements strings with number-based access to its characters are implementation details, not fundamental features.

> Although let me say again that, while I really understand @davidchambers concerns and yours about types, I firmly believe that primitives and their object wrappers should be treated as the same thing.

Here I disagree.  I am not interested in trying to see these the same.  The wrapper objects are a headache to my mind, but if someone uses them there is (one would hope) a real reason, and it's likely because of the differences in behavior they have with their companion primitives.  I don't know if we can actually come up with a reasonable implementation for `keysIn` that would do this, but this is what I think would be great:

```js
const x = 'foo'
x.bar = 'baz';
R.keys(x); //=> []

const y = new String('foo');
y.bar = 'baz';
R.keys(y); //=> ['bar']
```
This is in keeping with the differences I [noted in a comment][co] several years ago.


>>  As to isArrayLike, I think that's just nonsense that we should remove.

> My 2 cents: I think you should.
> Array-like is a concept, not a type and not even an interface.

However we deal with such concepts fairly often, including the data type we work with most often.  Ramda chooses to think of its sequence type as a _List_, even though there is not built-in List type.  We gloss over certain array details and try to work around others in order to treat these as List-like as we can.  But that adds something useful to Ramda.  I don't think `isArrayLike` does.

> [ ... ]
> I'm already sorry for transforming an issue tracker into a forum.

Not at all.  We tend to use our issues this way.  It actually _is_ what we use as a forum.

  [co]: https://github.com/ramda/ramda/pull/296#issuecomment-53722605

**Comment by ascartabelli** [comment]
@CrossEye:

>> Always thumbs up when your goal is consistency.

> Ok, but I'm afraid my reason for this change is different from yours. For me it's simply that keys/keysIn is about the names of object properties. And a string does not have properties named '0', '1', '2', etc, no more than a list does. 

My only reason is consistency: `keys` and `keysIn` would in fact gain nothing in usefulness by returning indexes as I'd prefer them to do.
As long as their behaviour is clear and documented, and consistent across the whole library, I'm happy.

> The facts that the language implements arrays atop objects and that it implements strings with number-based access to its characters are implementation details, not fundamental features.

Maybe not fundamental features, but I won't call them merely "implementation details" as they are exposed to the user.

One might not like it, but:

```javascript
[1, 2, 3].hasOwnProperty("0") // => true
"foo".hasOwnProperty("0") // => true
new String("foo").hasOwnProperty("0") // => true
```

>> I firmly believe that primitives and their object wrappers should be treated as the same thing.

> Here I disagree. I am not interested in trying to see these the same. The wrapper objects are a headache to my mind, but if someone uses them there is (one would hope) a real reason, and it's likely because of the differences in behavior they have with their companion primitives.

Correct me if I'm wrong, but the only differences that come to mind are:

1. they yield different results with the `typeof` operator, but we all have a `type` function that circumvents that
2. wrappers are extensible
3. ```new String("foo") === "foo" // => false```
4. they yield different results if evaluated with `eval`, e.g.:
```javascript
eval("2+2") // => 4
eval(new String("2+2")) // => String {"2+2", 0: "2", 1: "+", 2: "2", length: 3, __proto__: String} (Firefox console output)
```

So, you say:

>  I don't know if we can actually come up with a reasonable implementation for keysIn that would do this, but this is what I think would be great:
```javascript
const x = 'foo'
x.bar = 'baz';
R.keys(x); //=> []
const y = new String('foo');
y.bar = 'baz';
R.keys(y); //=> ['bar']
```

Should be easily doable, but is it worth the hassle?
Are you sure you don't need a check like that elsewhere to keep consistency?
And, more important, you are trying to defend yourself against what exactly?

Switching to my point of view, I:

1. use, as we all have it, a `type` function that yields the same result in both cases
2. do nothing, not my concern
3. use `Object` or `String` or `valueOf` depending on the case when I deal with values coming from the outside
4. do nothing, not my concern

So, unless I missed something, from where I stand it's like you're trying to defend yourself from grass attacking you, only to "defend the honor" of types, or your idea of how things should be implemented.
This is only silly irony, not trying to be sarcastic or polemic at all.

We may not like that ```"foo".hasOwnProperty("0")``` or that arrays are object, etc., but it's the language.
JavaScript is very versatile and, as Sanctuary did, you can build up a more strict environment: as I've said before it's a design choice I respect, but my point here wasn't about being strict regarding types.

My point is that I'm "worried" that you're trying to hide or circumvent some characteristics, or quirks, you don't like without a real benefit; and to achieve that you have to add code, checks, tests and probably minor performance hits and inconsistencies.

>> Array-like is a concept, not a type and not even an interface.

> However we deal with such concepts fairly often, including the data type we work with most often. Ramda chooses to think of its sequence type as a List, even though there is not built-in List type. We gloss over certain array details and try to work around others in order to treat these as List-like as we can. But that adds something useful to Ramda.

Sure, but are there checks for that in your codebase? Or are you simply "using things"?

**Comment by ascartabelli** [comment]
> Sure, but are there checks for that in your codebase? Or are you simply "using things"?

By the way, it wasn't rhetorical: I was really asking as I'm not familiar with your codebase.

**Comment by CrossEye** [comment]
I didn't mean to drop this.  I've been having problems with my internet provider and doing most of my entries on a tablet, not the best way to type.  Somehow I lost a long entry I started the other day in response to this and haven't had the heart to try retyping it.

**Comment by ascartabelli** [comment]
@CrossEye, don't worry: we can resume later if you like.

You can also try an abstract of your thoughts as an exercise in epigrammatic style, while I think of something that will require hippopotomonstrosesquipedaliophobia combined with some Welsh word as an answer.

Ok, no, maybe later is for the best: good luck with your provider, been there way too many times.

**Comment by asaf-romano** [comment]
It's only because of this issue that I now know that ``Object.keys`` works that way for strings. I wouldn't have guessed.

I always assumed that ``R.keys`` is pretty much the same as ``Object.keys``. Interestingly, R.keys does "dispatch" to to ``Object.keys``, but before doing so it checks that the input is an object. If it's not, it always returns an empty array. I personally wouldn't mind just dispatching to Object.keys.

**Comment by CrossEye** [comment]
@ascartabelli:

> We may not like that "foo".hasOwnProperty("0") or that arrays are object, etc., but it's the language.  JavaScript is very versatile

I simply don't feel any need to cater to every quirk of the language.  If people pass sparse arrays to Ramda functions that are looking for lists, bad things could happen.  Ramda might eat your hard drive, crash your car, steal your boyfriend, launch the nuclear weapons.  I won't try to prevent it.  Ramda thinks sometimes in terms of higher abstractions and ignores these details when they get in the way.  It is not trying to be a library that does everything for everyone.  There are already several good ones designed for that.  Ramda is promoting a particular style of work.  While that style should work with the grain of the language, it doesn't have to account for every knot in the wood.  To me the wrappers for the primitive objects are of that sort.  I understand why they exist, but I will not go out of my way to make anything work with them.


@asaf-romano: I wouldn't like to make this change and then having the function throw exceptions when supplied `null` or `undefined`.  I think that's the rationale around the current behavior.

**Comment by ascartabelli** [comment]
@CrossEye:
> I simply don't feel any need to cater to every quirk of the language. If people pass sparse arrays to Ramda functions that are looking for lists, bad things could happen. Ramda might eat your hard drive, crash your car, steal your boyfriend, launch the nuclear weapons. I won't try to prevent it. Ramda thinks sometimes in terms of higher abstractions and ignores these details when they get in the way.

On this I agree completely, although the very same reasoning makes me act in a very different way.
Probably it's because I consider things like `"foo".hasOwnProperty("0") === true` as inherent characteristics of the language rather than quirks.

Anyway, even without discussing what can be classified as quirk, I stand by my point about raising excessive defenses against nothing.

> I wouldn't like to make this change and then having the function throw exceptions when supplied `null` or `undefined`

Well, this would be easily solved, but what's your rationale behind the decision?

I thought about it a lot and, unsurprisingly, decided to stick with the language and have exceptions with `nil` values.
At first my main doubt was about "hiding too much" versus "hassling the user", but then I shifted my point of view and started thinking in terms of problem domain.
In a nutshell my reasoning is that I'm only offering a basic toolset to let users build higher abstractions; hence I should stick with the language behaviour and let such decisions be in the domain created by the user of the tools.

**Comment by CrossEye** [comment]
>> I wouldn't like to make this change and then having [_who proofread this? - ed._] the function throw exceptions when supplied null or undefined

> Well, this would be easily solved, but what's your rationale behind the decision?

I'm not sure how easy this would be to solve by "just dispatching to Object.keys", but the rationale is that I prefer functions that are simple and composable.  I certainly am willing to make concessions to the language, but they should be rare.

Deciding not to throw is a pretty straightforward decision for me:  When I want to compose functions, I really need to understand their behavior.  `keys` accepts an object and returns a list of strings.  That's pretty simple.  [Postel's Law][pl] insists that we remain cautious about expanding on our output here.  If as well as returning a list of strings it also throws an exception, then we've expanded the surface area a consumer has to match.  And we make it much harder to include in a composition.

But expanding it to also accept nil values doesn't hurt me any.  I can still return a list of strings; it would simply have to be an empty list.  That's a lot easier to deal with on the output side.

  [pl]: https://en.wikipedia.org/wiki/Robustness_principle

**Comment by ascartabelli** [comment]
>>> I wouldn't like to make this change and then having [who proofread this? - ed.] the function throw exceptions when supplied null or undefined

>> Well, this would be easily solved, but what's your rationale behind the decision?

> I'm not sure how easy this would be to solve by "just dispatching to Object.keys"

@CrossEye, I was referring to what I proposed in the first post, when I thought this behaviour was a bug.

> Deciding not to throw is a pretty straightforward decision for me

Well, it wasn't for me doing the opposite and it still isn't as I feel a bit conflicted, actually, but I think I chose for the best in my case.
Anyway as long as I'm behind the "1.0" mark, I won't feel too guilty about changing my mind later.
But I won't change my mind. Maybe. Pun intended.

Thanks for explaining your point of view: proper feedback, even on simple things, is a rare commodity.

**Comment by ascartabelli** [comment]
@CrossEye, sorry if I insist on the subject but, if you don't mind, can you please elaborate further?
Today I was looking at your REPL, because I like it **a lot**, and occurred to me that the behaviour of `R.keys` and `R.keysIn` seems to be the exception rather than the rule, as most of your functions throw when supplied with `nil` values in place of the "data" parameter.
I remember I tried all the "prop\<something\>" functions, and `map`, `filter` and `reduce` among others.

**Comment by CrossEye** [comment]
I'm not sure we've ever detailed our philosophy here, but it's been controversial enough that it lead to an entirely separate (and wonderful) library in [Sanctuary][sa].

@buzzdecafe and I started Ramda as a learning exercise and not as an attempt to make a library for the world.  Neither of use was ever interested in tools that did a lot of hand-holding.  We both were of the philosophy that if you supplied garbage, you deserved the garbage you got back.  And so neither of us ever tried to do much verification of user inputs.  But we also objected to functions that intentionally threw errors in any but the most dire circumstances, as those do not compose well at all.  (The very first controversy we had in Ramda was [around this issue][ft].  Although I started out arguing the other side of it, I eventually removed functions from Ramda rather than keep in ones that were intentionally throwing exceptions when they didn't have any other good options.)

Ramda is simply not as strict as Sanctuary.  And this is very intentional.  It's supposed to be a bit closer to the day-to-day lives of regular Javascript developers than to that of Haskell, OCaml, or Clojure developers.

For instance, here are a few calls to the first functions alphabetically in Ramda:

```js
add(10, 32);                     //=> 42
add('a', undefined);             //=> NaN
adjust(add(5), 10, [0, 1, 2]);   //=> [0, 1, 2]
adjust(add(5), 1, null);         // THROWS TypeError
adjust(add(5), 1, [0, 1, 2]);    //=> [0, 6, 2]
adjust(add(5), 1, {foo: 'bar'}); //=> [undefined, NaN]  -- WTF?
all(gt(__, 10), []);             //=> true
all(gt(__, 10), [12, 13, 14]);   //=> true
all(gt(__, 10), [9, 11, 13]);    //=> false
all(gt(__, 10), 'abc');          //=> false
all(gt(__, 10), {foo: 'bar'});   //=> true
all(gt(__, 10), null);           // THROWS TypeError
```

A number of these inputs don't make sense, and almost certainly if there are equivalents in Sanctuary, the run-time type-checking would have caught them and thrown useful exceptions.  Ramda walks a somewhat different path.

Generally, when we can give a reasonable answer even in the face of data that's not up to snuff, we try to do so.  For instance,

```js
propOr('Ramda', 'favLib', {});   //=> 'Ramda'
propOr('Ramda', 'favLib', null); //=> 'Ramda'
```

Because the user did not give a value in the second case, it might be reasonable to throw an exception, but we already know the expected value, so we can still answer reasonably.

I haven't looked carefully at all the cases you mention, but I think we might want to at least consider fixing `reduce`, as there should be a clear answer if the Foldable is not supplied.  `keys(null)` seems like the same sort of case to me.


  [ft]: https://github.com/ramda/ramda/issues/20
  [sa]: http://sanctuary.js.org/

**Comment by ascartabelli** [comment]
@CrossEye, I understand what you mean by saying "reasonable answer" and I thought myself of having some exceptions on the "nil rule", but more on that in a bit.
I'm worried that "reasonable" isn't good enough if you aren't in a null safe environment: when you have some functions throwing and some other making assumptions, things become confusing fast, no matter how much you are diligent in documenting behaviour.

In a null safe environment things are a bit better in that regard, but I think that the risk of burying a bug is too high and that the user should be the one deciding when to use `Maybe` or some safe version of the tool he's using.

So, as I said, while targeting ES5 I decided to stick with ES6 behaviour and let exceptions being raised for `nil` values and treat other unexpected types as the expected one; `Object.keys` being the perfect example:

```javascript
Object.keys({a: 1, b: 2}) // => ["a", "b"]
Object.keys(2) // => [] (would have thrown a TypeError in ES5)
Object.keys(null) // => throws a TypeError
```

So there's some form of "reasonable answer", but no null safety.

I mentioned thinking about exceptions to the rule and a `merge` function comes to mind: my `merge` now behaves exactly like Ramda's, mine's is \[still?\] variadic though (oh, this could lead to another interesting discussion).
The closest thing we have in ES6 is `Object.assign` but, obviously, we don't have a target object being mutated.
The absence of the ugly target also means the we don't have one main "data parameter" so, in my case, I'm still unsure if I should consider all the received objects as "main data" or not.
Probably if I can contain such exceptions to a very limited but coherent set of cases I'll be happy nonetheless.

As a *post scriptum*, let me clarify that all of this philosophical dullness is about API design: I'm way more relaxed in everyday coding.

**Comment by CrossEye** [comment]
@ascartabelli: To me there is a difference in whether we can comfortably answer the question being asked by the function invocation.  This one

```js
keys({a: 1, b: 2});
```

asks the question, "What are the (enumerable, own) string keys of the object `{a: 1, b: 2}`?  And of course we can answer that one easily enough, with `['a', 'b']`.  This one

```js
keys(null);
```

asks, "What are the (enumerable, own) string keys of `null`?  Well, it doesn't have any.  But that's easy enough to express in our output format, with `[]`.

Similarly, this one:

```js
adjust(add(5), 1, [0, 1, 2]);
```

asks the question, "What is the result of adjusting the second parameter (i.e., the one at index 1) of the list `[0, 1, 2]` with the function `add(5)`?"  The answer is clear, and the code can handle this without issue, returning `[0, 6, 2]`.  But this one:

```js
adjust(add(5), 1, null);
```

asks, "What is the result of adjusting the second parameter (i.e., the one at index 1) of `null` with the function `add(5)`?  We have no fallback.  Here's where we end up throwing.  Often we don't do so intentionally in Ramda; it simply falls out of the implementation; there are actually very, very few `throw` clauses in Ramda.

Note that this is not specifically to do with the "data" parameter being null.  The same thing would happen if the first parameter were not a function.  And you'd get nonsense -- although due to type coercion, not an exception -- if the second parameter were not an integer.

> As a *post scriptum*, let me clarify that all of this philosophical dullness is about API design: I'm way more relaxed in everyday coding.

I do find this sort of API discussion worth having, and for me it does bleed a bit into everyday coding, I'm afraid.  :smile:

**Comment by ascartabelli** [comment]
@CrossEye, I'm a bit in a hurry right now but I see an interesting point of view: should have time to elaborate properly tomorrow.
In the meantime kudos to you for not biting at my variadic bait.

> I do find this sort of API discussion worth having

Me too, and it's good to have good company. The more the merrier.

> and for me it does bleed a bit into everyday coding, I'm afraid.

Yes, it really does, but at least in private code you can, say, write a function, specify a type you require in a doc comment, commit and think about edge cases in a more relaxed way.

**Comment by ascartabelli** [comment]
@CrossEye, it's my turn now to say that I didn't mean to drop this: sorry but I got sidetracked by life.
Going back on the matter at hand soon, though.

**Comment by CrossEye** [comment]
No problem.  I'm not exactly working at lightning speed myself these days.

**Comment by ascartabelli** [comment]
@CrossEye: ok, I really didn't mean to drop this, but I almost forgot.

Your point of view about a "reasonable answer" is interesting and makes sense, though I'm not sure it can be rationalized to a proper rule.
It seems a bit too questionable for a matter where I feel the need of something more strict.
I do think it's still valid as a conceptual approach.

Speaking of rules, just out of curiosity, I recently broke one of mine.

When I said:

> One of the design goals of my library is not throwing exceptions at all explicitly: I have to be able to code "letting things flow" and let the language itself be the one throwing them.

I used that admittedly very odd "goal" as a complexity meter; telling me that if I find myself throwing exceptions manually it probably meant that I stepped over an hypothetical line and made something too complex for my intended scope.

Well, I was wrong. 
In the process of streamlining the API's behaviour, I promptly found a bunch of very simple cases where I was forced to throw explicitly.
